transport/awssqs/consumer.go

```diff
@@ -0,0 +1,305 @@
+	for _, f := range c.after {
+		ctx = f(ctx, msg, &responseMsg, leftMsgs, c.leftMsgsMux)
+	}
+
+	if !c.wantRep(ctx, msg) {
+		return nil
+	}
+
+	if err := c.enc(ctx, &responseMsg, response); err != nil {
+		c.errorHandler.Handle(ctx, err)
+		c.errorEncoder(ctx, err, msg, c.sqsClient)
+		return err
+	}
+
+	if _, err := c.sqsClient.SendMessageWithContext(ctx, &responseMsg); err != nil {
+		c.errorHandler.Handle(ctx, err)
+		c.errorEncoder(ctx, err, msg, c.sqsClient)
+		return err
+	}
+	return nil
+}
+
+// ErrorEncoder is responsible for encoding an error to the consumer's reply.
+// Users are encouraged to use custom ErrorEncoders to encode errors to
+// their replies, and will likely want to pass and check for their own error
+// types.
+type ErrorEncoder func(ctx context.Context, err error, req *sqs.Message, sqsClient Client)
+
+// ConsumerFinalizerFunc can be used to perform work at the end of a request
+// from a publisher, after the response has been written to the publisher. The
+// principal intended use is for request logging.
+// Can also be used to delete messages once fully proccessed.
+type ConsumerFinalizerFunc func(ctx context.Context, msg *[]*sqs.Message)
+
+// VisibilityTimeoutFunc encapsulates logic to extend messages visibility timeout.
+// this can be used to provide custom visibility timeout extension such as doubling it everytime
+// it gets close to being reached.
+// VisibilityTimeoutFunc will need to check that the provided context is not done and return once it is.
+type VisibilityTimeoutFunc func(context.Context, Client, string, int64, *[]*sqs.Message, *sync.Mutex) error
+
+// WantReplyFunc encapsulates logic to check whether message awaits response or not
+// for example check for a given message attribute value.
+type WantReplyFunc func(context.Context, *sqs.Message) bool
+
+// DefaultErrorEncoder simply ignores the message. It does not reply.
+func DefaultErrorEncoder(context.Context, error, *sqs.Message, Client) {
+}
+
+// DoNotExtendVisibilityTimeout is the default value for the consumer's visibilityTimeoutFunc.
+// It returns no error and does nothing
+func DoNotExtendVisibilityTimeout(context.Context, Client, string, int64, *[]*sqs.Message, *sync.Mutex) error {
+	return nil
+}
+
+// DoNotRespond is a WantReplyFunc and is the default value for consumer's wantRep field.
+// It indicates that the message do not expect a response.
+func DoNotRespond(context.Context, *sqs.Message) bool {
+	return false
+}
+
+// EncodeJSONResponse marshals response as json and loads it into an sqs.SendMessageInput MessageBody.
+func EncodeJSONResponse(_ context.Context, input *sqs.SendMessageInput, response interface{}) error {
+	payload, err := json.Marshal(response)
+	if err != nil {
+		return err
+	}
+	input.MessageBody = aws.String(string(payload))
+	return nil
+}
+
+// Client is consumer contract for the Producer and Consumer.
+// It models methods of the AWS *sqs.SQS type.
+type Client interface {
```

Comment:
https://github.com/aws/aws-sdk-go/blob/master/service/sqs/sqsiface/interface.go

SQS already has their API interface

Reply:
Thank you for your input. I just made the switch.